apicast/src/oauth/apicast_oauth.lua

```diff
@@ -1,26 +1,417 @@
+
+  return params
+end
+
 function _M.transform_credentials(_, credentials)
   return credentials
 end
 
-return _M
+function _M.respond_and_exit(status, body, headers)
+  -- TODO: is there a better way to populate the response headers?..
+  if headers then
+    for name,value in pairs(headers) do
+      ngx.header[name] = value
+    end
+  end
+
+  ngx.status = status
+  ngx.print(body)
+  ngx.exit(ngx.HTTP_OK)
+end
+
+function _M.respond_with_error(status, message, description)
+
+  --TODO: as per the RFC (https://tools.ietf.org/html/rfc6749#section-5.2) return WWW-Authenticate response header if 401
+  local headers = {
+    ['Content-Type'] = 'application/json;charset=UTF-8'
+  }
+  local err_msg = { error = message, error_description = description }
+  local body = cjson.encode(err_msg)
+  ngx.log(ngx.INFO, "error :" .. inspect(body))
+  _M.respond_and_exit(status, body, headers)
+end
+
+-- TODO: Split error conditions up further to decide when we should respond with error and when we should redirect_with error
+function _M.redirect_with_error(url, error, state)
+  ngx.header.content_type = "application/x-www-form-urlencoded"
+  return ngx.redirect(url,"?error=",error.error,"&error_description=",error.error_description,"&state=",state)
+end
+
+function _M.check_params(params)
+  local response_type = params.response_type
+  local required_params = _M.params.response_type
+  if not response_type then return false, 'invalid_request' end
+  if not required_params[response_type] then return false, 'unsupported_response_type' end
+
+  for _,v in ipairs(required_params[response_type]) do
+    if not params[v] then
+      return false, 'invalid_request'
+    end
+  end
+
+  return true
+end
+
+function _M.token_check_params(params)
+  local grant_type = params.grant_type
+  local required_params = _M.params.grant_type
+  if not grant_type then return false, 'invalid_request' end
+  if not required_params[grant_type] then return false, 'unsupported_grant_type' end
+
+  for _,v in ipairs(required_params[grant_type]) do
+    if not params[v] then
+      return false, 'invalid_request'
+    end
+  end
+  return true
+end
+
+function _M.check_credentials(service, params)
+  local backend = backend_client:new(service)
+
+  local args = {
+      app_id = params.client_id,
+      app_key = params.client_secret,
+      redirect_uri = params.redirect_uri
+    }
+
+  local res = backend:authorize(args)
+
+  return res.status == 200
+end
+
+-- returns a unique string for the client_id. it will be short lived
+local function nonce(client_id)
+  return ts.sha1_digest(tostring(random.bytes(20, true)) .. "#login:" .. client_id)
+end
+
+local function generate_access_token(client_id)
+  local token = ts.sha1_digest(tostring(random.bytes(20, true)) .. client_id)
+
+  return { ["access_token"] = token, ["token_type"] = "bearer", ["expires_in"] = env.get('APICAST_OAUTH_ACCESS_TOKEN_TTL') or 604800 }
```

Comment:
no need to use string keys, those are all valid lua table identifiers without quotes